File: crates/snipx-core/src/ast.rs
use rowan::NodeOrToken;

use crate::diagnostic::SourceSpan;
use crate::syntax::{SyntaxKind, SyntaxNode};

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct DirectiveValue {
    pub value: String,
    pub span: SourceSpan,
}

#[derive(Debug, Clone, Default, PartialEq, Eq)]
pub struct HeaderDirectives {
    pub profile: Option<DirectiveValue>,
    pub target: Option<DirectiveValue>,
    pub duplicates: Vec<(String, SourceSpan)>,
}

pub trait AstNode: Sized {
    fn can_cast(kind: SyntaxKind) -> bool;
    fn cast(node: SyntaxNode) -> Option<Self>;
    fn syntax(&self) -> &SyntaxNode;

    fn text(&self) -> String {
        self.syntax().to_string()
    }
}

macro_rules! ast_node {
    ($name:ident, $($kind:pat_param)|+ $(,)?) => {
        #[derive(Debug, Clone, PartialEq, Eq)]
        pub struct $name {
            syntax: SyntaxNode,
        }

        impl AstNode for $name {
            fn can_cast(kind: SyntaxKind) -> bool {
                matches!(kind, $($kind)|+)
            }

            fn cast(node: SyntaxNode) -> Option<Self> {
                Self::can_cast(node.kind()).then_some(Self { syntax: node })
            }

            fn syntax(&self) -> &SyntaxNode {
                &self.syntax
            }
        }
    };
}

ast_node!(Root, SyntaxKind::Root);
ast_node!(
    SnipxRegion,
    SyntaxKind::Root | SyntaxKind::LineComment | SyntaxKind::Fence | SyntaxKind::IntralineaBlock
);
ast_node!(Statement, SyntaxKind::Statement);
ast_node!(Subject, SyntaxKind::Subject);
ast_node!(Predicate, SyntaxKind::Predicate);
ast_node!(Object, SyntaxKind::Object);
ast_node!(ObjectList, SyntaxKind::ObjectList);
ast_node!(Snippet, SyntaxKind::Snippet);
ast_node!(RangeSnippet, SyntaxKind::RangeSnippet);
ast_node!(Capture, SyntaxKind::Capture);
ast_node!(
    Directive,
    SyntaxKind::Directive | SyntaxKind::TargetDirective | SyntaxKind::ProfileDirective
);
ast_node!(Decoration, SyntaxKind::Decoration);
ast_node!(IntralineaBlock, SyntaxKind::IntralineaBlock);

impl Root {
    pub fn statements(&self) -> impl Iterator<Item = Statement> + '_ {
        self.syntax.descendants().filter_map(Statement::cast)
    }

    pub fn regions(&self) -> Vec<SnipxRegion> {
        let child_regions: Vec<_> = self
            .syntax
            .children()
            .filter_map(SnipxRegion::cast)
            .filter(SnipxRegion::contains_snipx_items)
            .collect();

        if !child_regions.is_empty() {
            child_regions
        } else if SnipxRegion::contains_snipx_items_for(&self.syntax) {
            vec![SnipxRegion {
                syntax: self.syntax.clone(),
            }]
        } else {
            Vec::new()
        }
    }

    pub fn directives(&self) -> impl Iterator<Item = Directive> + '_ {
        self.syntax.descendants().filter_map(Directive::cast)
    }

    pub fn header_directives(&self) -> HeaderDirectives {
        let mut header = HeaderDirectives::default();
        for directive in self.directives() {
            let slot = match directive.syntax().kind() {
                SyntaxKind::ProfileDirective => ("profile", &mut header.profile),
                SyntaxKind::TargetDirective => ("target", &mut header.target),
                _ => continue,
            };
            let Some(value) = directive.value() else {
                continue;
            };
            let (name, slot) = slot;
            if slot.is_none() {
                *slot = Some(value);
            } else {
                header.duplicates.push((name.to_owned(), value.span));
            }
        }
        header
    }
}

impl SnipxRegion {
    pub fn statements(&self) -> impl Iterator<Item = Statement> + '_ {
        self.syntax.children().filter_map(Statement::cast)
    }

    pub fn directives(&self) -> impl Iterator<Item = Directive> + '_ {
        self.syntax.children().filter_map(Directive::cast)
    }

    pub fn intralinea_blocks(&self) -> impl Iterator<Item = IntralineaBlock> + '_ {
        self.syntax.children().filter_map(IntralineaBlock::cast)
    }

    fn contains_snipx_items(&self) -> bool {
        Self::contains_snipx_items_for(&self.syntax)
    }

    fn contains_snipx_items_for(node: &SyntaxNode) -> bool {
        node.children().any(|child| {
            matches!(
                child.kind(),
                SyntaxKind::Statement
                    | SyntaxKind::Directive
                    | SyntaxKind::TargetDirective
                    | SyntaxKind::ProfileDirective
                    | SyntaxKind::IntralineaBlock
            )
        })
    }
}

impl Statement {
    pub fn subject(&self) -> Option<Subject> {
        child(&self.syntax)
    }

    pub fn predicates(&self) -> impl Iterator<Item = Predicate> + '_ {
        self.syntax.children().filter_map(Predicate::cast)
    }

    pub fn predicate(&self) -> Option<Predicate> {
        self.predicates().next()
    }

    pub fn object_lists(&self) -> impl Iterator<Item = ObjectList> + '_ {
        self.syntax.children().filter_map(ObjectList::cast)
    }

    pub fn object_list(&self) -> Option<ObjectList> {
        self.object_lists().next()
    }

    pub fn decorations(&self) -> impl Iterator<Item = Decoration> + '_ {
        self.syntax.children().filter_map(Decoration::cast)
    }
}

impl Subject {
    pub fn snippets(&self) -> impl Iterator<Item = Snippet> + '_ {
        self.syntax.descendants().filter_map(Snippet::cast)
    }

    pub fn range_snippets(&self) -> impl Iterator<Item = RangeSnippet> + '_ {
        self.syntax.descendants().filter_map(RangeSnippet::cast)
    }
}

impl ObjectList {
    pub fn objects(&self) -> impl Iterator<Item = Object> + '_ {
        self.syntax.children().filter_map(Object::cast)
    }
}

impl Object {
    pub fn decorations(&self) -> impl Iterator<Item = Decoration> {
        let decorations = self
            .syntax
            .parent()
            .filter(|parent| parent.kind() == SyntaxKind::ObjectList)
            .map(|parent| {
                let mut seen_object = false;
                let mut decorations = Vec::new();

                for child in parent.children() {
                    if !seen_object {
                        seen_object = child == self.syntax;
                        continue;
                    }

                    if child.kind() == SyntaxKind::Object {
                        break;
                    }
                    if let Some(decoration) = Decoration::cast(child) {
                        decorations.push(decoration);
                    }
                }

                decorations
            })
            .unwrap_or_default();

        decorations.into_iter()
    }
}

impl Directive {
    pub fn name(&self) -> Option<String> {
        self.syntax
            .children_with_tokens()
            .find_map(|element| match element {
                NodeOrToken::Node(node) if node.kind() == SyntaxKind::Identifier => {
                    Some(node.to_string())
                }
                NodeOrToken::Token(token) if token.kind() == SyntaxKind::Identifier => {
                    Some(token.text().to_string())
                }
                _ => None,
            })
    }
}

impl Directive {
    /// The directive's value: the URI body for `@target`, or the first
    /// value identifier for `@profile`. The directive name itself is an
    /// identifier token, not a node, so it never matches here.
    pub fn value(&self) -> Option<DirectiveValue> {
        self.syntax.children().find_map(|child| match child.kind() {
            SyntaxKind::Uri => {
                let text = child.to_string();
                let value = text
                    .strip_prefix('<')
                    .and_then(|text| text.strip_suffix('>'))
                    .unwrap_or(&text)
                    .to_owned();
                Some(DirectiveValue {
                    value,
                    span: node_span(&child),
                })
            }
            SyntaxKind::Identifier => Some(DirectiveValue {
                value: child.to_string(),
                span: node_span(&child),
            }),
            _ => None,
        })
    }
}

impl Snippet {
    pub fn captures(&self) -> impl Iterator<Item = Capture> + '_ {
        self.syntax.descendants().filter_map(Capture::cast)
    }
}

fn node_span(node: &SyntaxNode) -> SourceSpan {
    let range = node.text_range();
    SourceSpan {
        start: u32::from(range.start()) as usize,
        end: u32::from(range.end()) as usize,
    }
}

fn child<N: AstNode>(node: &SyntaxNode) -> Option<N> {
    node.children().find_map(N::cast)
}

File: packages/snipx-ts/src/ast.ts
/** Mirror of the AST accessors in crates/snipx-core/src/ast.rs. */

import type { SourceSpan } from "./diagnostic.js";
import type { SyntaxNode, SyntaxToken } from "./syntax.js";

export interface DirectiveValue {
  value: string;
  /** UTF-16 code-unit span into the source (converted at export). */
  span: SourceSpan;
}

export interface HeaderDirectives {
  profile: DirectiveValue | null;
  target: DirectiveValue | null;
  duplicates: [string, SourceSpan][];
}

export function nodeSpan(node: SyntaxNode): SourceSpan {
  return { start: node.start, end: node.end };
}

export function statementSubject(statement: SyntaxNode): SyntaxNode | null {
  return statement.childNodes().find((child) => child.kind === "Subject") ?? null;
}

export function statementPredicates(statement: SyntaxNode): SyntaxNode[] {
  return statement.childNodes().filter((child) => child.kind === "Predicate");
}

export function statementObjectLists(statement: SyntaxNode): SyntaxNode[] {
  return statement.childNodes().filter((child) => child.kind === "ObjectList");
}

export function statementDecorations(statement: SyntaxNode): SyntaxNode[] {
  return statement.childNodes().filter((child) => child.kind === "Decoration");
}

export function objectListObjects(list: SyntaxNode): SyntaxNode[] {
  return list.childNodes().filter((child) => child.kind === "Object");
}

/**
 * Decorations attached to an object: the Decoration siblings after it in
 * the ObjectList, up to the next Object.
 */
export function objectDecorations(object: SyntaxNode): SyntaxNode[] {
  const parent = object.parent;
  if (parent === null || parent.kind !== "ObjectList") {
    return [];
  }
  let seenObject = false;
  const decorations: SyntaxNode[] = [];
  for (const child of parent.childNodes()) {
    if (!seenObject) {
      seenObject = child === object;
      continue;
    }
    if (child.kind === "Object") {
      break;
    }
    if (child.kind === "Decoration") {
      decorations.push(child);
    }
  }
  return decorations;
}

export function isDirectiveKind(node: SyntaxNode): boolean {
  return (
    node.kind === "Directive" || node.kind === "TargetDirective" || node.kind === "ProfileDirective"
  );
}

/**
 * The directive's value: the URI body for `@target`, or the first value
 * identifier for `@profile`. The directive name itself is an identifier
 * token, not a node, so it never matches here.
 */
export function directiveValue(directive: SyntaxNode): DirectiveValue | null {
  for (const child of directive.childNodes()) {
    if (child.kind === "Uri") {
      const text = child.toText();
      const stripped =
        text.startsWith("<") && text.endsWith(">") && text.length >= 2
          ? text.slice(1, -1)
          : text;
      return { value: stripped, span: nodeSpan(child) };
    }
    if (child.kind === "Identifier") {
      return { value: child.toText(), span: nodeSpan(child) };
    }
  }
  return null;
}

export function headerDirectives(root: SyntaxNode): HeaderDirectives {
  const header: HeaderDirectives = { profile: null, target: null, duplicates: [] };
  for (const node of root.descendants()) {
    if (!isDirectiveKind(node)) continue;
    let name: "profile" | "target";
    if (node.kind === "ProfileDirective") {
      name = "profile";
    } else if (node.kind === "TargetDirective") {
      name = "target";
    } else {
      continue;
    }
    const value = directiveValue(node);
    if (value === null) {
      continue;
    }
    if (header[name] === null) {
      header[name] = value;
    } else {
      header.duplicates.push([name, value.span]);
    }
  }
  return header;
}

export function tokensOf(node: SyntaxNode): SyntaxToken[] {
  const tokens: SyntaxToken[] = [];
  for (const element of node.descendantsWithTokens()) {
    if (element.type === "token") {
      tokens.push(element);
    }
  }
  return tokens;
}
